File: exams/tipantiza/u3/OOP30812FinalExamExercise/StrategyPattern/src/main/java/ec/edu/espe/strategypattern/model/InsertionSort.java

package ec.edu.espe.strategypattern.model;

/**
 *
 * @author Alexander Tipantiza, CodeBreakers, @ESPE
 */

public class InsertionSort implements SortingStrategy {
    @Override
    public int[] sort(int[] numbers) {
        int[] array = numbers.clone();
        for (int i = 1; i < array.length; i++) {
            int key = array[i];
            int j = i - 1;
            while (j >= 0 && array[j] > key) {
                array[j + 1] = array[j];
                j = j - 1;
            }
            array[j + 1] = key;
        }
        return array;
    }
}

File: exams/tipan/u3/Exam U3/StrategyJS/model/InsertionSort.js
const SortingStrategy = require("./SortingStrategy");

class InsertionSort extends SortingStrategy {

    sort(array) {
        let result = [...array];

        for (let i = 1; i < result.length; i++) {

            let current = result[i];
            let j = i - 1;

            while (j >= 0 && result[j] > current) {
                result[j + 1] = result[j];
                j--;
            }

            result[j + 1] = current;
        }

        return result;
    }
}

module.exports = InsertionSort;